File: 74 Rock Scissors Paper/python/rockscissors.py
#!/usr/bin/env python3
# ROCKSCISSORS
#
# Converted from BASIC to Python by Trevor Hobson

import random


def play_game():
    """Play one round of the game"""

    while True:
        try:
            games = int(input("How many games? "))
            if games >= 11:
                print("Sorry, but we aren't allowed to play that many.")
            else:
                break

        except ValueError:
            print("Please enter a number.")

    won_computer = 0
    won_human = 0

    for game in range(games):
        print("\nGame number", game + 1)
        guess_computer = random.randint(1, 3)
        print("3=Rock...2=Scissors...1=Paper")
        guess_human = 0
        while guess_human == 0:
            try:
                guess_human = int(input("1...2...3...What's your choice "))
                if guess_human not in [1, 2, 3]:
                    guess_human = 0
                    print("Invalid")

            except ValueError:
                print("Please enter a number.")
        print("This is my choice...")
        if guess_computer == 1:
            print("...Paper")
        elif guess_computer == 2:
            print("...Scissors")
        elif guess_computer == 3:
            print("...Rock")
        if guess_computer == guess_human:
            print("Tie Game. No winner")
        elif guess_computer > guess_human:
            if guess_human != 1 or guess_computer != 3:
                print("Wow! I win!!!")
                won_computer = won_computer + 1
            else:
                print("You win!!!")
                won_human = won_human + 1
        elif guess_computer == 1:
            if guess_human != 3:
                print("You win!!!")
                won_human = won_human + 1
            else:
                print("Wow! I win!!!")
                won_computer = won_computer + 1
    print("\nHere is the final game score:")
    print("I have won", won_computer, "game(s).")
    print("You have won", won_human, "game(s).")
    print("and", games - (won_computer + won_human), "game(s) ended in a tie.")
    print("\nThanks for playing!!\n")


def main():
    print(" " * 21 + "GAME OF ROCK, SCISSORS, PAPER")
    print(" " * 15 + "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY\n")

    keep_playing = True
    while keep_playing:

        play_game()

        keep_playing = input(
            "Play again? (yes or no) ").lower().startswith("y")


if __name__ == "__main__":
    main()

File: 00_Alternate_Languages/74_Rock_Scissors_Paper/nim/rockscissors.nim
import std/[random,strformat,strutils]

type
  symbol = enum
    PAPER = 1, SCISSORS = 2, ROCK = 3

var
  cpuChoice, playerChoice, turns: int
  cpuWins, playerWins, ties: int = 0

randomize()

# Function: player makes a choice
proc choose(): int =
  echo "3=ROCK...2=SCISSORS...1=PAPER...WHAT'S YOUR CHOICE?"
  result = readLine(stdin).parseInt()

# Function: determine the outcome
proc outcome(p: symbol, c: symbol): string =
  if p == c:
    ties += 1
    result = "TIE GAME.  NO WINNER."
  else:
    const
      winTable = [
        PAPER: (ROCK, "COVERS"),
        SCISSORS: (PAPER, "CUTS"),
        ROCK: (SCISSORS, "CRUSHES")
      ]
    let (winCond, winVerb) = winTable[p]
    if winCond == c:
      playerWins += 1
      result = fmt"{p} {winVerb} {c}.  YOU WIN."
    else:
      let (_, winVerb) = winTable[c]
      cpuWins += 1
      result = fmt"{c} {winVerb} {p}.  I WIN."

# Start the game
echo spaces(21), "GAME OF ROCK, SCISSORS, PAPER"
echo spaces(15), "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY"
echo "\n"
echo "HOW MANY GAMES?"
turns = readLine(stdin).parseInt()
while turns > 10:
  echo "SORRY, BUT WE AREN'T ALLOWED TO PLAY THAT MANY."
  turns = readLine(stdin).parseInt()

# Play the game
for i in 1..turns:
  echo ""
  echo "GAME NUMBER ", i
  playerChoice = choose()
  while playerChoice != 1 and playerChoice != 2 and playerChoice != 3:
    echo "INVALID"
    playerChoice = choose()
  cpuChoice = rand(1..3) # match against range in symbol
  echo "THIS IS MY CHOICE... ", symbol(cpuChoice)
  echo outcome(symbol(playerChoice), symbol(cpuChoice))

# Results
echo ""
echo "HERE IS THE FINAL GAME SCORE:"
echo "I HAVE WON ", cpuWins," GAME(S)."
echo "YOU HAVE WON ", playerWins," GAME(S)."
echo "AND ", ties," GAME(S) ENDED IN A TIE."
echo ""
echo "THANKS FOR PLAYING!!"
